cloudflare/resource_cloudflare_rate_limit.go

```diff
@@ -56,12 +56,12 @@ func resourceCloudflareRateLimit() *schema.Resource {
 						"mode": {
 							Type:         schema.TypeString,
 							Required:     true,
-							ValidateFunc: validation.StringInSlice([]string{"simulate", "ban"}, true),
+							ValidateFunc: validation.StringInSlice([]string{"simulate", "ban", "challenge", "js_challenge"}, true),
 						},
 
 						"timeout": {
 							Type:         schema.TypeInt,
-							Required:     true,
```

Comment:
As this is now optional based on the mode, we probably want to add a check in the `Create`/`Update` that if the mode is simulate or ban, we check the value is defined to prevent people scratching their heads with unintended behaviour.

Reply:
I completely agree. The Cloudflare API does give reasonable error messages around this though it should really be caught by the provider before the API is called.

I'm just working on it at the moment, though this is the first time I've written any Go so you may have to forgive my non-idiomatic code.

Reply:
After grappling with attempting to insert validation into `expandRateLimitAction`, I've realised that this won't behave quite as expected. I've attempted to implement a `ValidateFunc` into the `action` section of the schema itself (where I'll have access to both `mode` and `timeout`), but it appears that this is not supported on lists or sets at this point in time.

I'd appreciate any pointers as to how to proceed with this.

Reply:
> ... but it appears that this is not supported on lists or sets at this point in time.

Yep, the `ValidateFunc` is only supported on primitive types which throws it out a bit. 

> I'd appreciate any pointers as to how to proceed with this.

Sure, I _think_ this might do the job. You were on the right track with using `expandRateLimitAction` as it's used in both the `Create` and `Update`. I haven't thoroughly vetted this though. 

```diff
diff --git a/cloudflare/resource_cloudflare_rate_limit.go b/cloudflare/resource_cloudflare_rate_limit.go
index e67f1a4..dd0f170 100644
--- a/cloudflare/resource_cloudflare_rate_limit.go
+++ b/cloudflare/resource_cloudflare_rate_limit.go
@@ -197,11 +197,15 @@ func resourceCloudflareRateLimit() *schema.Resource {
 
 func resourceCloudflareRateLimitCreate(d *schema.ResourceData, meta interface{}) error {
 	client := meta.(*cloudflare.API)
+	rateLimitAction, err := expandRateLimitAction(d)
+	if err != nil {
+		return errors.Wrap(err, "error expanding rate limit action")
+	}
 
 	newRateLimit := cloudflare.RateLimit{
 		Threshold: d.Get("threshold").(int),
 		Period:    d.Get("period").(int),
-		Action:    expandRateLimitAction(d),
+		Action:    rateLimitAction,
 	}
 
 	newRateLimitMatch, err := expandRateLimitTrafficMatcher(d)
@@ -255,11 +259,15 @@ func resourceCloudflareRateLimitUpdate(d *schema.ResourceData, meta interface{})
 	client := meta.(*cloudflare.API)
 	zoneId := d.Get("zone_id").(string)
 	rateLimitId := d.Id()
+	rateLimitAction, err := expandRateLimitAction(d)
+	if err != nil {
+		return errors.Wrap(err, "error expanding rate limit action")
+	}
 
 	updatedRateLimit := cloudflare.RateLimit{
 		Threshold: d.Get("threshold").(int),
 		Period:    d.Get("period").(int),
-		Action:    expandRateLimitAction(d),
+		Action:    rateLimitAction,
 	}
 
 	newRateLimitMatch, err := expandRateLimitTrafficMatcher(d)
@@ -341,13 +349,21 @@ func expandRateLimitTrafficMatcher(d *schema.ResourceData) (matcher cloudflare.R
 	return
 }
 
-func expandRateLimitAction(d *schema.ResourceData) cloudflare.RateLimitAction {
+func expandRateLimitAction(d *schema.ResourceData) (cloudflare.RateLimitAction, error) {
 	// dont need to guard for array length because MinItems is set **and** action is required
 	tfAction := d.Get("action").([]interface{})[0].(map[string]interface{})
+	mode := tfAction["mode"].(string)
+	timeout := tfAction["timeout"].(int)
+
+	if mode == "simulate" || mode == "ban" {
+		if timeout == 0 {
+			return cloudflare.RateLimitAction{}, fmt.Errorf("rate limit timeout must be set if the 'mode' is simulate or ban")
+		}
+	}
 
 	action := cloudflare.RateLimitAction{
-		Mode:    tfAction["mode"].(string),
-		Timeout: tfAction["timeout"].(int),
+		Mode:    mode,
+		Timeout: timeout,
 	}
 
 	if _, ok := tfAction["response"]; ok && len(tfAction["response"].([]interface{})) > 0 {
@@ -359,7 +375,7 @@ func expandRateLimitAction(d *schema.ResourceData) cloudflare.RateLimitAction {
 			Body:        tfActionResponse["body"].(string),
 		}
 	}
-	return action
+	return action, nil
 }
 
 func expandRateLimitCorrelate(d *schema.ResourceData) (correlate cloudflare.RateLimitCorrelate, err error) {
```

Do you want to try this and let me know how you go?

Reply:
That's absolutely brilliant, thanks for your help. I'd used `fmt.Errorf` to raise errors, but hadn't wrapped them with `errors.Wrap`. This now raises errors correctly.

I expanded the logic to throw an error when timeout values are provided for `challenge` and `js_challenge` modes as the Cloudflare API won't accept them.